### PR TITLE
Throw TypeError when accessing array with out-of-bound index in strict mode

### PR DESCRIPTION
**Pull request by ThomsonTan**
This is for issue https://github.com/Microsoft/ChakraCore/issues/2083

### CODE REVIEW

**Comment by digitalinfinity** [comment]
@dotnet-bot test this please

**Comment by ThomsonTan** [comment]
@jianchun could you please help take a look?

File: lib/Runtime/Types/ES5ArrayTypeHandler.cpp [hunk 1]
```diff
@@ -371,7 +371,7 @@ namespace Js
         // Reject if we need to grow non-writable length
         if (!CanSetItemAt(arr, index))
         {
```
Comment (jianchun) [inline lib/Runtime/Types/ES5ArrayTypeHandler.cpp]: There are 2 more similar fragments in this file ("if (!CanSetItemAt...) { return false; }"), please double check if they need fixes too.

Comment (ThomsonTan) [inline lib/Runtime/Types/ES5ArrayTypeHandler.cpp]: @jianchun I checked some others cases like Object.defineProperty which doesn't enter the other occurrences in the code. How about check in this with a comment? Probably extend it later.

**Comment by ThomsonTan** [comment]
Talked with @jianchun offline and it is Ok to take the fix now.

**Pull request merged**